tests/riscv64/deeplabv3_float_rvv_test.run

```diff
@@ -0,0 +1,13 @@
+// RUN: ${TEST_MODULE_CMD} \
+// RUN:   --module_file="${BUILD_RISCV_DIR}/tests/e2e/models/tflite_llvm-cpu_local-task_deeplabv3.tflite_module.vmfb" \
+// RUN:   --entry_function=main --function_input="1x257x257x3xf32=0" --print_max_element_count=1008 | \
+// RUN:   grep "=\[" | tee %t
+//
+// RUN: ${HOST_TEST_MODULE_CMD} \
+// RUN:   --module_file="${BUILD_RISCV_DIR}/tests/e2e/models/tflite_llvm-cpu_local-task_x86_64_deeplabv3.tflite_module.vmfb" \
+// RUN:   --entry_function=main --function_input="1x257x257x3xf32=0" --print_max_element_count=1008 | \
+// RUN:   grep "=\[" | tee %t.x86_64
+//
+// Compare numerical results.
+// RUN: python ${IREE_ROOT_DIR}/build_tools/testing/compare_run_module_output.py \
+// RUN:   --tol=0.001 %t %t.x86_64
```

Comment:
Again - let's not add backend specific tests for whole program behavior in the core project. These need to run across all backends for us to have any hope of being able to maintain and debug them.

Reply:
I think running across all backends is different then running across all cpu arches; the former can be done in host machine as long as it supports GPU/CPU/etc, where as in crosscompile a different set of flags needs to be supported and a device/emulator needs to be provided.

Back to this particular lit test. If we don't want to build tosa frontend full program in `iree_check_test`, another possibility in CI would be fetching the tarball from `build_benchmarks` step and run the lit test with the hardcoded expected values. The goal here is not to maintain and debug LLVM ***in IREE codebase***, but to provide signals when we update llvm-project so we can debug/fix RISCV flow upstream.

I have trouble figuring out the best way to support #9824 as well as @dcaballe 's needs of having more numerical correctness checks. As discussed offline, I agree `iree_check_test` is _NOT_ the best place to generate the workload by hooking up the network as well as `iree-import-tflite` , but where should it be? We already discussed the challenge/infeasiblity of using `iree-samples` to support this for cross compile targets, so how do we want to proceed?

The alternative is to somehow enable the actual benchmark CI on the RISCV platform, but I am not sure it is going to be available soon, or if anyone is working on it.

Reply:
I think all the comment threads are on the same basic issue, so just commenting here.

We should be separating benchmarks from correctness testing here and the ordering should definitely be correctness testing before benchmarking. The benchmarks currently make use of iree-import-tflite and fetch from the network. This is bad, but notably the benchmarks are *not* the core project (and we've even been discussing moving them into a separate CMake project or even repository).

I think what Scott is saying is that we shouldn't be testing anything *only* on one unusual backend, where the backend here is "llvm to RISC-V codegen". The first step for any test should be making sure it runs on our simplest and most-supported configurations:  x86-64 Linux with LLVM CPU and VMVX. From there we can build out coverage to other backends.

Starting with e2e tests might be appropriate for a "spike" or "prototype" to test out some new thing, but that's for experimental work where the person doing the development is making personal tradeoffs around development velocity/correctness/completeness/etc. So something in an experimental folder might have limited test coverage to prove out some path as remotely feasible, but before it becomes non-experimental it needs to be running unit tests. For instance, Scott's work on wasm and WebGPU runtime is in the experimental directory and doesn't affect any CI. Because it's not bothering anyone else, he may have decided to start with some e2e prototypes (I have no idea if this was the case), but now we're talking about making it non-experimental, so he's started an RFC for how we're going to make all this work (and like RISC-V it's not going to be trivial). I know I'm going to insist that it runs our unit tests before we start adding integration tests to the CI. The RISC-V stuff is really in that experimental state now wrt to its test coverage. That's not a criticism: I totally get that there are some unique challenges to working on a CPU architecture that basically doesn't exist in actual silicon yet. And we already have some CI for it that mostly makes sure that everything builds (which is good), but before we ask the team at large to take on general responsibility for this working, we need to build up the foundation of testing starting with unit tests.

If we do need more e2e model tests after that, we should be adding them to our more supported and easier to debug backends before (or at the same time as) adding them to more esoteric ones.

There are all kinds of logistical hurdles here and we've made tradeoffs for expediency before, but that doesn't mean we can just skip all the initial steps and jump straight to the end here.

Bringing things back to the concrete, I see a few things we're trying to tackle:
1. Running all our runtime unit tests on RISCV-64. Can we do this the same way we do for Android now where there's some setup and wrapper around the command? That's still not awesome, but it's workable and I see a natural evolution from that to a more generic "test running configuration" similar to Bazel's `--run_under`.
2. A desire to have more full-model tests. I remain skeptical that this is *actually* especially necessary for RISC-V as opposed to papering over other issues. Let's assume that it is something we want regardless though. We've been hampered by the lack of a stable interchange format below the frontend level. I'm not sure there's a great solution till we have MLIR bytecode + stable input dialect. I think the former is pretty much ready to go, and hopefully StableHLO will give us the latter, though I'm not clear whether we intend to make that IREE's input dialect. In the meantime, we're kind of stuck with TFLite Flatbuffers, which unfortunately brings in a TF dependency :-/ I think we can be creative about how we limit the impact of that, probably by getting this test coverage via the integrations/ build until we get proper stable interchange formats.

Let's continue discussing the goals here and how to resolve them, but I don't think this is a case of "just fix this one thing in the PR". I understand you're trying to get something done here, but we really need to take one thing at a time.

Reply:
Thank you Geoffrey for the detailed feedback. I understand we all try to do the right thing; at the same time, RISCV llvm has given us so much fun and surprise, so if possible, we would like to catch issues missed upstream to provide some safety net. Not necessarily mean we need to fix the bug, but at least we can catch the breakage during the presubmit.

Regarding your two suggested moves.

1. Add RV64 runtime unit tests: That's kind of straightforward. Here is a preview of what I got so far
```
96% tests passed, 4 tests failed out of 92

Label Time Summary:
driver=local-sync                  =   0.81 sec*proc (16 tests)
driver=local-task                  =   2.92 sec*proc (26 tests)
iree/base                          =   0.95 sec*proc (6 tests)
iree/base/internal                 =   1.92 sec*proc (12 tests)
iree/base/testing                  =   0.05 sec*proc (1 test)
iree/builtins/device/tools         =   0.11 sec*proc (2 tests)
iree/builtins/ukernel/tools        =   0.10 sec*proc (2 tests)
iree/hal                           =   0.08 sec*proc (1 test)
iree/hal/drivers/local_sync/cts    =   0.81 sec*proc (16 tests)
iree/hal/drivers/local_task/cts    =   2.56 sec*proc (22 tests)
iree/hal/drivers/vulkan/util       =   0.14 sec*proc (3 tests)
iree/hal/local                     =   0.08 sec*proc (2 tests)
iree/hal/local/elf                 =   0.02 sec*proc (1 test)
iree/hal/utils                     =   0.16 sec*proc (3 tests)
iree/modules/check                 =   0.14 sec*proc (1 test)
iree/modules/check/test            =   0.36 sec*proc (4 tests)
iree/task                          =   2.16 sec*proc (8 tests)
iree/tooling                       =   0.10 sec*proc (2 tests)
iree/vm                            =   0.27 sec*proc (6 tests)
noasan                             =   0.04 sec*proc (1 test)

Total Test time (real) =  10.06 sec

The following tests FAILED:
	 10 - iree/base/internal/file_io_test (Failed)
	 12 - iree/base/internal/fpu_state_test (Failed)
	 19 - iree/base/testing/dynamic_library_test (Failed)
	 99 - iree/tooling/numpy_io_test (Subprocess aborted)
```
2. Some creative ways of getting more full-model tests. I can grab the results from `build_benchmarks` step and conditionally add the test in `build_tools/cmake/test_riscv64.sh`. That might be the least invasive way. WDYT?

Reply:
Thank you all for the engagement and discussion! I agree with Cindy that we all are trying to do the right thing here :). I think we need to jump on a meeting to address this more effectively. Let me schedule one for next week and add a few more people that can help. 

A few quick replies/clarifications:
* Running unit tests on RISC-V is a must, absolutely! I think we all agree on this. However, unit tests don't address the problems we are trying to solve here: 1) as Cindy mentioned above, we need a frontline defense against the LLVM RISC-V backend, which is actively under development and, 2) we are also dealing with a product reputation issue that I'll elaborate in the meeting. We need integration and correctness testing to address them.
* I do agree that testing portability is highly desirable! We should work towards that goal. However, I don't think we can assume that all the backends will run the same tests. Different targets have different goals and capabilities (compute, memory, hw emulation :) ). Mobile or embedded targets may not be able to run server caliber integration tests/models, for example, or an extensive test suite. I think we need a bit more flexibility here. Adding tests to “more supported” targets shouldn't be a requirement for introducing a new test in our infra for “less supported” targets. 
* I think it's important that our infra is able to properly support testing scenarios like the RISC-V one (emulation) as they are a fundamental part of our HW/SW co-design strategy and goals.

I'll schedule a meeting.